arelle/packages/report/ReportPackage.py

```diff
@@ -30,28 +31,75 @@ def _getReportPackageTopLevelDirectory(filesource: FileSource) -> str | None:
-    packageJsonPath = f"{stld}/{Const.REPORT_PACKAGE_FILE}"
-    if stld is None or packageJsonPath not in (filesource.dir or []):
+def getReportPackageJsonFile(filesource: FileSource, stld: str | None) -> str | None:
+    if not isinstance(filesource.fs, zipfile.ZipFile):
+        return None
+    # Match against original unsanitized file names and all their backslash glory.
+    packageOriginalFileNames = [f.orig_filename for f in filesource.fs.infolist()]
+    expectedFutureReportPackageJsonPath = Const.REPORT_PACKAGE_FILE
+    expectedReportPackageJsonPath = f"{stld}/{Const.REPORT_PACKAGE_FILE}"
+    futureReportPackageJsonPathPartialMatches = []
+    reportPackageJsonPathPartialMatches = []
+    for path in packageOriginalFileNames:
+        if path.startswith(expectedFutureReportPackageJsonPath):
+            futureReportPackageJsonPathPartialMatches.append(path)
+        elif path.startswith(expectedReportPackageJsonPath):
+            reportPackageJsonPathPartialMatches.append(path)
+
+    jsonPath = None
+    # Don't return content if there are duplicate entries or a directory with the same name as the json file.
+    if len(futureReportPackageJsonPathPartialMatches) == 1:
+        futureReportPackageJsonPath = futureReportPackageJsonPathPartialMatches[0]
+        if futureReportPackageJsonPath == expectedFutureReportPackageJsonPath:
+            jsonPath = expectedFutureReportPackageJsonPath
+    elif len(futureReportPackageJsonPathPartialMatches) == 0:
+        if len(reportPackageJsonPathPartialMatches) == 1:
+            reportPackageJsonPath = reportPackageJsonPathPartialMatches[0]
+            if reportPackageJsonPath == expectedReportPackageJsonPath:
+                jsonPath = expectedReportPackageJsonPath
+    return jsonPath
+
+
+def getPackageJson(filesource: FileSource, selection: str | None) -> dict[str, Any] | None:
+    if selection is None:
         return None
     initialSelection = filesource.selection
     initialUrl = filesource.url
-    try:
-        filesource.select(packageJsonPath)
-        fullPackageJsonPath = cast(str, filesource.url)
-        with filesource.file(fullPackageJsonPath, binary=True)[0] as rpj:
-            return cast(dict[str, Any], json.load(rpj))
-    except (OSError, zipfile.BadZipFile, json.JSONDecodeError):
-        return None
-    finally:
-        filesource.selection = initialSelection
-        filesource.url = initialUrl
+    filesource.select(selection)
+    fullJsonFilePath = cast(str, filesource.url)
+    content = None
+    encodings = ["utf-8", "utf-8-sig"]
+    for encoding in encodings:
+        try:
+            with filesource.file(fullJsonFilePath, encoding=encoding)[0] as f:
+                content = cast(dict[str, Any], json.load(f, object_pairs_hook=forbidDuplicateKeys))
+                break
+        except (OSError, ValueError, zipfile.BadZipFile, json.JSONDecodeError):
+            continue
```

Comment:
If we catch+ignore all errors, where do we report errors in the file?  Is there another code path where we do validations rather than loading the data?  If so, that's a little unfortunate, but maybe just a comment would be helpful.

Reply:
I pushed an update to log the error (same pattern is used in the `filesource.open()` method), but validation is deferred to validation classes that have the full context of which specifications are being validated against and which errors should be raised.